docs/proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,121 @@
+
+### 1. Current use of Cloud Provider
+
+The following components have cloudprovider dependencies
+
+    1. kube-controller-manager
+    2. kubelet
+    3. kube-apiserver
+
+#### Cloud Provider in Kube-Controller-Manager
+
+The kube-controller-manager has many controller loops
+
+ - nodeController
+ - volumeController
+ - routeController
+ - serviceController
+ - replicationController
+ - endpointController
+ - resourceQuotaController
+ - namespaceController
+ - deploymentController
+ - etc..
+
+Among these controller loops, the following are cloud provider dependent.
+
+ - nodeController
+ - volumeController
+ - routeController
+ - serviceController
+
+The nodeController uses the cloudprovider to check if a node has been deleted from the cloud. If cloud provider reports a node as deleted, then this controller immediately deletes the node from kubernetes. This check removes the need to wait for a specific amount of time to conclude that an inactive node is actually dead.
+
+The volumeController uses the cloudprovider to create, delete, attach and detach volumes to nodes. For eg, the logic for provisioning, attaching, and detaching a EBS volume resides in the AWS cloudprovider. The volumeController uses this code to perform its operations.
+
+The routeController configures routes for hosts in Google Cloud.
+
+The serviceController maintains a list of currently active nodes, and is responsible for creating and deleting LoadBalancers in the underlying cloud.
+
+#### Cloud Provider in Kubelet
+
+Moving on to the kubelet, the following cloud provider dependencies exist in kubelet.
+
+ - Find the cloud nodename of the host that kubelet is running on for the following reasons :
+      1. To obtain the config map for the kubelet, if one already exists
+      2. To uniquely identify current node using nodeInformer
+      3. To instantiate a reference to the current node object
+ - Find the InstanceID, ProviderID, ExternalID, Zone Info of the node object while initailizing it
+ - Periodically poll the cloud provider to figure out if the node has any new IP addresses associated with it
+ - In case of Google cloud, it allows the provider to configure routes for the node
+ - It allows the cloud provider to post process DNS settings
+
+#### Cloud Provider in Kube-apiserver
+
+Finally, in the kube-apiserver, the cloud provider is used for trasferring SSH keys to all of the nodes, and within an admission controller for setting labels on persistent volumes.
+
+### 2. Strategy for refactoring Kube-Controller-Manager
+
+In order to create a 100% cloud independent controller manager, we will split the controller-manager into multiple binaries.
+
+1. Cloud dependent controller-manager binaries
```

Comment:
Maybe point out that this is the "normal" `controller-manager` that's shipped with the Kubernetes releases